File: backend/apps/authentication/validators/identification_validator.py
"""Type-aware Ecuadorian identification validation for account data."""

from __future__ import annotations

from apps.authentication.models import User
from core.base.base_validator import BaseValidator, ValidationResult


class IdentificationValidator(BaseValidator):
    """Require exactly 13 RUC digits or 10 cédula digits."""

    LENGTH_BY_TYPE = {
        User.TipoIdentificacion.RUC: 13,
        User.TipoIdentificacion.CEDULA: 10,
    }

    def validate(self, data: dict) -> ValidationResult:
        identification_type = data.get(
            "tipo_identificacion", User.TipoIdentificacion.RUC
        )
        identification = data.get("ruc")

        if identification_type not in self.LENGTH_BY_TYPE:
            return ValidationResult(
                is_valid=False,
                errors=["Selecciona un tipo de identificación válido."],
                field_name="tipo_identificacion",
            )
        if not isinstance(identification, str) or not identification:
            return ValidationResult(
                is_valid=False,
                errors=["La identificación es obligatoria."],
                field_name="ruc",
            )
        if not identification.isascii() or not identification.isdigit():
            return ValidationResult(
                is_valid=False,
                errors=[
                    "La identificación debe contener solo dígitos, sin espacios ni guiones."
                ],
                field_name="ruc",
            )

        expected_length = self.LENGTH_BY_TYPE[identification_type]
        if len(identification) != expected_length:
            label = "El RUC" if identification_type == User.TipoIdentificacion.RUC else "La cédula"
            return ValidationResult(
                is_valid=False,
                errors=[f"{label} debe tener exactamente {expected_length} dígitos."],
                field_name="ruc",
            )
        return ValidationResult(is_valid=True)

File: frontend/src/modules/auth/validators/IdentificationValidator.ts
/** Type-aware validation for RUC and cédula account identification. */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'
import type { IdentificationType } from '../interfaces/IAuthService'

interface IdentificationData {
  tipoIdentificacion?: IdentificationType
  ruc?: string
}

export const IDENTIFICATION_LENGTH: Record<IdentificationType, number> = {
  RUC: 13,
  Cedula: 10,
}

export class IdentificationValidator extends BaseValidator {
  validate(data: unknown): ValidationResult {
    const { tipoIdentificacion = 'RUC', ruc = '' } = (data ?? {}) as IdentificationData
    const expectedLength = IDENTIFICATION_LENGTH[tipoIdentificacion]

    if (!expectedLength) {
      return {
        isValid: false,
        errors: ['Selecciona un tipo de identificación válido.'],
        field: 'tipoIdentificacion',
      }
    }
    if (!ruc) {
      return { isValid: false, errors: ['La identificación es obligatoria.'], field: 'ruc' }
    }
    if (!/^\d+$/.test(ruc)) {
      return {
        isValid: false,
        errors: ['La identificación debe contener solo dígitos, sin espacios ni guiones.'],
        field: 'ruc',
      }
    }
    if (ruc.length !== expectedLength) {
      const label = tipoIdentificacion === 'RUC' ? 'El RUC' : 'La cédula'
      return {
        isValid: false,
        errors: [`${label} debe tener exactamente ${expectedLength} dígitos.`],
        field: 'ruc',
      }
    }
    return { isValid: true, errors: [], field: 'ruc' }
  }
}

export function hasCompleteIdentification(data: IdentificationData): boolean {
  return new IdentificationValidator().validate(data).isValid
}
